### DISCUSSION TITLE
Test levels - verification of commands

### DISCUSSION THREAD

**Discussion by mefistotelis** [swfans/syndwarsfx#231] [Ideas]
Currently only few simple unit tests of `bflibrary` are executed during the builds. These verify if math functions work, or whether drawing polygons works exactly as expected (images are generated and [compared to reference](https://github.com/swfans/swars/tree/master/bflibrary/tests/referenc)).

We do have some tests for the whole game as well - [swars-tests](https://github.com/swfans/swars-tests) contains a collection of packet files which can be played to just load a level, let it play for a few seconds, then exit. So these can test loading a level an progressing the world. We will sooner or later add executing these to the automated CI tests, it's just a matter of writing a script.

But these only test elements which happen relatively early at start of a level, and it's more like crash detection than proper verification of what happens within the game world.

What we need I think is a set of levels for verifying specific elements of the game. And the ones we should verify first are probably Commands.

So, as soon as we have a level editor functional, it would be good to start working on a level for verifying commands. It's best to select a map with a lot of features, so that more commands can be checked - ie. boarding a train, opening a dome, catching a ferry (that one we expect to not work currently), moving to places on 3D objects rather than ground.

Besides verification, having such level would also help in understanding how each command works.

**Comment by Moburma** [comment]
I can do all that now! What map do we think is best? I would say probably either Bangkok or Cairo, they both have trains and domes, and both even have water.

Ferries are interesting, but I think would be a ton of work for little reward in getting them going. There seem to be two problems:
a) stating the obvious, you can't get inside them (I tested this loads before, you just can't do it)
b) They're not "real" vehicles in the sense other ones are. Aside from not being able to get inside them, they can't have drivers or commands. They just wallhug the piece of water they are in, turning away from any wall they hit to try and keep going in a straight line. This makes them unusable as an actual ferry vehicle as you can't make them just go from a to b as would be required. They're really just a visual effect that only works in a handful of levels.

**Reply by mefistotelis** [reply]
> They just wallhug the piece of water they are in, turning away from any wall they hit to try and keep going in a straight line. 

This isn't actually that bad, we just need to make them stop for some time near docks/ports.
The commands which are not used anywhere are actually easy to fix, because we can change how they work without consequences on the game. Like, I don't like some commands using absolute face number as go-to point - the game never uses that, and I don't blame the mapmakers for it, as even a slight change in any building will change the face number. It was a bad idea to use face numbers at all. But since they're unused, we will probably just change the commands - to either require a face index within a specific building, or skip that at all - the game can figure out the best face during level loading.
(example for the above - NPC boarding a train requires a command with face number on which to stand on the station floor)

Anyway back to the subject, yes we should start the work on such level. Cairo has docks and usable bridge objects, so it's probably better. 
There is no pair of docks, but if I remember right there's only one map in the whole game which has more of them. I also can't remember Cairo having any domes.. but that might be my memory issue. If we can't make everything happen on one map, we can create two levels, this shouldn't be an issue as well.

I'm still wondering how to make verification whether everything worked. What I expect is that after some amount of turns - like maybe 500 - the level will get terminated. At this point a check will be made on all the people on map, and that will determine the outcome.

The check should be simple, ie. whether a person finished all assigned commands (so ComCur == 0), or maybe also got some flag set (maybe IgnoreEnemies). It must be something which won't happen if the person did not fulfilled the commands, and can be used for all people on the test level.

I guess the simplest solution would be a "wait until" command which checks position of the person. Such command can be placed at end of chain for every person. So if the person is at position we expect, the condition will be met immediately setting ComCur to 0. If it won't be met - the command is "wait" so it will never reach the destination marking the error by having unfulfilled command.

**Reply by Moburma** [reply]
>  It was a bad idea to use face numbers at all. 

When I was saw this a while ago I just thought wtf, I don't know how you'd even get that working right outside of basically the actual game in an editor, not sure if the old map reader has that information.

>  I also can't remember Cairo having any domes.. but that might be my memory issue. 

Yeah I'm getting it wrong, it's actually Bahrain I'm thinking of - has rail, domes and water.

> I guess the simplest solution would be a "wait until" command which checks position of the person. Such command can be placed at end of chain for every person. So if the person is at position we expect, the condition will be met immediately setting ComCur to 0. If it won't be met - the command is "wait" so it will never reach the destination marking the error by having unfulfilled command.

That's a good idea, I will have a go at it tomorrow.

**Reply by Moburma** [reply]
I had a quick bash at this, annoyingly I've been much busier lately than I hoped to be. It's a pretty simple level, although I realised after  I started this map does Not actually have any trains after all..

https://easyupload.io/fayhbc

Looks like Destroy_building actually doesn't work? Looking through other levels they just use use_weapon with explosives.

I will put way more tests in, this is just for now.

**Reply by mefistotelis** [reply]
> Looks like Destroy_building actually doesn't work?

Well, that's good - we've just found a first bug. The person tries to move to a position with high accuracy, but its speed is too high. So it continues to step over the target place.

**Comment by mefistotelis** [comment]
The closer I look, the more test cases I see. Even for that single command.
* what if a person is in a vehicle?
* What about a vehicle with own weapon, like a tank or mech?
* what about different weapons? explosive? nuclear grenade? launcher?
* if a person has multiple weapons in inventory, will it select the best one for the job?
* building walkable inside vs building without doors, will they both work?

Anyway, I did some updates to the game. The bomb is now deployed.

EDIT:
The bomb is now planted, but then the person just stands there. Nothing more is implemented.

I guess we need some general plan for this command.. maybe:
1. Go near target point, use weapon
2. if weapon is a bomb, run away somewhere (where?) and wait for explosion
3. after explosion, check if building destroyed. yes -> finish state; no -> repeat from point 1

**Reply by Moburma** [reply]
Ahh awesome, I can see it's working now. It is an interesting command, I guess the extra complexity wasn't needed in the end as it was simpler to just employee Use-Weapon. As you say potentially they could be attacking a building with almost any weapon, but I'm not sure it's worth expanding that much as nothing uses it and it's a little redundant. 

Also I would think you would just script in the rest using other commands? Otherwise you're maybe assuming too much in regards to how a level designer would want them to behave - maybe they'd WANT them to do a suicide bombing!

I'm not very busy today, so will get some more tests created.

**Comment by Moburma** [comment]
Think I found a map bug:
https://easyupload.io/xsjcpt

I added a vehicle on the road leading to the bridge. Look what happens when it tries to cross the bridge. I assume this is why none of the levels on this map have cars actively driving about...

**Comment by mefistotelis** [comment]
> I added a vehicle on the road leading to the bridge. Look what happens when it tries to cross the bridge.

There are no four-lane roads, so each road has still two directions. This one is probably not a bug, just lack of implementation of any other road that with opposite directions.

When it reaches the map end, the car joins the party at (0,0). There are already some things there, hard to distinguish as they all have the exact same position.

**Reply by Moburma** [reply]
It's a bit more fundamental than that:
https://www.youtube.com/watch?v=Z7IKud2fBkU
It's always when it tries to cross the bridge there, sometimes it gets closer than others

**Reply by mefistotelis** [reply]
Try using thing debug hud and locking on the vehicle (`w`). I'm pretty sure it's just another end of the transport node which teleports the vehicle to (0,0).

EDIT:
Wait, it's not (0,0). The vehicle keeps going, just somewhere else. And only this specific lane causes the teleport.

Looks like I got confused by the off-map things on this level; the game can't seem to correctly compute distance near map borders. Will have to fix that.

**Reply by Moburma** [reply]
Yes, sorry, I should have explained that detail. Coming back the other way it's fine. I assume it's similar to the weird Adelaide bug where the cars teleport into the prison.

**Reply by mefistotelis** [reply]
Looks like someone was deleting something and grabbed a few traffic nodes with it. Not sure if they didn't used vehicles on this map because of it, or they just didn't noticed that simple deletion mistake because there were no vehicles. We'll see - maybe something more will get revealed when we try to fix that.

**Comment by Moburma** [comment]
Another odd command, PERSUADE_ALL_GROUP
https://easyupload.io/uk4slb

This has an issue that consistently happens where the persuader creeps after the target like they are in Protect mode, and they don't actually persuade them. Then suddenly after about 30 seconds it kicks in and they quickly persuade them all without a hitch. I had the same thing when I toyed with this for my Trash Church map. I forced persuadertron via SELECT_WEAPON here, but I'm not sure that is actually needed, it's the persuade command causing the weirdness.

**Reply by Moburma** [reply]
Another bit of persuasion testing.
https://easyupload.io/l03ljb
This is an interesting one I remember seeing before as well. PERSUADE_PERSON has the same weird pause before it kicks in and actually happens, but it has a further bug where the persuade graphical effect (where they kind of flash blue) never plays (the sound effect does, though).

**Reply by mefistotelis** [reply]
The weapon ranges, they seem to be computed differently in every place.
Will try to unify.

Grenades have longer range with high arms, but looks like that was not visible by cursors on target. Fixed that.
But persuader should have range increased by brain, right? Looks like this is also implemented in one place only, for now I didn't found the computation.

EDIT: is the persuadertron range based on brain mod just broken? I found some code, and it suggests higher skin mods have effect on the range rather than brain.. is the code that buggy or am I reading it wrong?
EDIT2: Ok, it's not that bad. The compiler just optimized the selection of bits in an unusual way... it should work based on brains. But - only player agents get bonuses from mods. For everyone else, mods are ignored.

**Reply by mefistotelis** [reply]
The problem with persuasion still persists! I mean the first person uses the range I set, so it is persuaded smoothly. but for GROUP commands, the following people have range reset somewhere else.. no idea where.

**Reply by mefistotelis** [reply]
Remade another function, and now fully fixed it. Though now I wonder.. did we just made the game harder this way?
It turns out for non-player characters the persuasion doesn't even cost weapon energy, so now the command will always be executed quite fast.

**Reply by Moburma** [reply]
To be fair, these commands are used so rarely I don't think it matters. I'm pretty sure there is no level-critical NPC persuade event in any mission. I think this thread is evidence of why that is! It's actually quite weird they didn't delve into this mechanic more, but the fact it worked so poorly implemented I guess shows why.

**Comment by Moburma** [comment]
Another version with some new things. Does AVOID_PERSON work now? bit hard to say here, I might move the testers. Also put in a woman with camera commands, just stand next to her by the water to see. Might add camera rotate as well.
https://easyupload.io/hkf2m5

**Reply by mefistotelis** [reply]
> Does AVOID_PERSON work now?

It works mostly as it did before. A bit better, but is still "run in opposite direction" rather than "wander but avoid".

I did written my thoughts on this in one of recent "issues".

**Comment by Moburma** [comment]
Added some more commands the other day:
https://easyupload.io/o65b21

Hide doesn't seem to do anything, or maybe I'm using it wrong, but then what _should_ it actually do? Things like this seem interesting to me, as with the likes of the early "WorkPlace" and "LeisurePlace" variables I think there was once an idea to make the game a much more complex open world simulation at its heart. The problem is it's just not needed for what is ultimately a fairly simple shoot em up game on very small maps.

**Reply by mefistotelis** [reply]
>  I think there was once an idea to make the game a much more complex open world simulation at its heart.

Yes, it kind of looks this way. The state machines of people were supposed to be more complex. Looking at the code, I have a feeling some programmers had still some remains of Amiga approach - I mean optimization rather than abstraction. This contrasts with Dungeon Keeper, where some basic abstraction layers were defined, and mostly kept.

Here, optimization is the king:
* thing position is stored and computed differently for every thing, just to avoid some additional bit shifting operations while rendering
* things have states and sub-states, but the sub-states are barely used and instead other conditions are checked for figuring out "real" substates
* Generic methods for searching all the things were limited and buggy, due to lack of abstraction

Adding anything more complex to this would require a rewrite of the entire state handling. So it feels like a learning ground for what was then implemented in DK - there, every creature goes somewhere with a purpose.

EDIT:
Have you noticed the NPCs never attack mounted lasers? That's because searching for targets happens on a list of things on tiles, for each tile in some surrounding area. But as an optimization, only some things are added to that list - people and vehicles, and only the ones "pinged" on map. So for the perception of NPCs, nothing exists besides people and vehicles. And mounted gun is a building.

**Reply by Moburma** [reply]
> some programmers had still some remains of Amiga approach

Yeah, that makes sense. Mike Diskett was hired to work on Amiga ports of their games, Syndicate Wars was the first (and sadly also last) original PC game he worked on there (and he did the Amiga port of Syndicate, which was 100% an optimisation task as other programmers at Bullfrog had tried and failed to do it). 

>Have you noticed the NPCs never attack mounted lasers?

Ahh, I hadn't, but now you mention it, yes.

**Reply by mefistotelis** [reply]
> Hide doesn't seem to do anything, or maybe I'm using it wrong, but then what should it actually do?

Current implementation of `HIDE` is just blank. For ferry, there are some chunks of code here and there; but `HIDE` doesn't even seem to have accompanying person state.

I guess the idea would be to run between walls of buildings, waiting for stamina between the runs. Then enter one of the buildings, and wait there until anyone with a gun enters, in which case select another building and repeat.

Maybe also avoiding cops while doing it? Or avoid a specified group? If we specify a group, maybe selecting new building should happen only if someone from that group enters?

We should discuss how each new command should work in a separate thread, otherwise this will get too messy to get any information.
